### PR TITLE
Adjustments to the parser behavior for a Simple Program

### PR DESCRIPTION
**Pull request by AlekseyTs**
- Follow up on parser changes that came from master
- Presence of attributes shouldn’t disable parsing of a top-level statement. Instead, the attributes should be attached to the top-level statement with an error reported, or without an error if the statement is a local function declaration.
- Allow a declaration of a `ref` local at the top level.

Related to #41704.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 1]
```diff
@@ -2356,6 +2371,30 @@ private MemberDeclarationSyntax ParseMemberDeclarationOrStatementCore(SyntaxKind
...
+                switch (statement?.Kind)
+                {
+                    case null:
+                    case SyntaxKind.LocalDeclarationStatement:
+                    case SyntaxKind.LocalFunctionStatement:
+                    case SyntaxKind.ExpressionStatement when
+                                !isScript &&
+                                // Do not parse a single identifier as an expression statement in a Simple Program, this could be a beginning of a keyword and
+                                // we want completion to offer it.
+                                ((ExpressionStatementSyntax)statement) is var exprStatement &&
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: It seems like a null 'statement' is matched here and then we go on to unsafely access members on it.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: > It seems like a null 'statement' is matched here and then we go on to unsafely access members on it.

See ```case null:``` above.

---
In reply to: [388021510](https://github.com/dotnet/roslyn/pull/42147#discussion_r388021510) [](ancestors = 388021510)

**Comment by AlekseyTs** [comment]
@cston, @RikkiGibson, @dotnet/roslyn-compiler Please review

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 2]
```diff
@@ -2021,15 +2027,19 @@ private MemberDeclarationSyntax ParseMemberDeclarationOrStatementCore(SyntaxKind
...
+
+            var afterAttributesPoint = this.GetResetPoint();
+
+            var modifiers = _pool.Allocate();
 
             try
             {
-                var attributes = this.ParseAttributeDeclarations();
-                bool haveAttributes = (attributes.Count > 0);
+                TerminatorState saveTerm;
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: >TerminatorState saveTerm [](start = 16, length = 24)

Please move inside nested `try`.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 3]
```diff
@@ -2109,7 +2118,7 @@ private MemberDeclarationSyntax ParseMemberDeclarationOrStatementCore(SyntaxKind
                     //            missing ';'
                     //
                     // Unless there modifiers or attributes are present this is more likely to be a method call than a method definition.
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: >there  [](start = 30, length = 6)

Not part of the PR, but consider correcting the sentence, perhaps by removing "there".

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 4]
```diff
@@ -2021,15 +2027,19 @@ private MemberDeclarationSyntax ParseMemberDeclarationOrStatementCore(SyntaxKind
...
+            var afterAttributesPoint = this.GetResetPoint();
+
+            var modifiers = _pool.Allocate();
 
             try
             {
-                var attributes = this.ParseAttributeDeclarations();
-                bool haveAttributes = (attributes.Count > 0);
+                TerminatorState saveTerm;
+                bool wasInAsync;
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: >bool wasInAsync [](start = 16, length = 15)

Please consider removing this local and keeping the individual locals in each scope, for clarity.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 5]
```diff
@@ -2177,30 +2186,30 @@ private MemberDeclarationSyntax ParseMemberDeclarationOrStatementCore(SyntaxKind
...
                             {
                                 IsInAsync = true; // We are implicitly in an async context
                             }
-                            // We don't allow local declaration statements at the top level.  We want
-                            // to fall out below and parse them instead as fields.
-                            var statement = this.ParseStatementCore(attributes, isGlobalScriptLevel: true);
-
+                            // In Script we don't allow local declaration statements at the top level.  We want
+                            // to fall out below and parse them instead as fields. For Simple Programs, we allow
+                            // them, but want to try  properties , etc. first.
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: >  properties  [](start = 52, length = 13)

Extra spaces.

**Comment by RikkiGibson** [comment]
>         public void CS1514ERR_LbraceExpected02()

this test seems like a copy-paste of the one in ParserErrorMessageTest, is it included here for ease of access?

---
Refers to: src/Compilers/CSharp/Test/Syntax/Parsing/SimpleProgramsParsingTests.cs:769 in ada365a. [](commit_id = ada365a432978d5e899faa2a8d1a4d760934973b, deletion_comment = False)

**Comment by RikkiGibson** [comment]
>         public void CS1514ERR_LbraceExpected02()

ah, seems like one validates the tree as well as parsing diagnostics and the other validates just the parsing diagnostics.

---
In reply to: [595432854](https://github.com/dotnet/roslyn/pull/42147#issuecomment-595432854) [](ancestors = 595432854)

---
Refers to: src/Compilers/CSharp/Test/Syntax/Parsing/SimpleProgramsParsingTests.cs:769 in ada365a. [](commit_id = ada365a432978d5e899faa2a8d1a4d760934973b, deletion_comment = False)

File: src/Compilers/CSharp/Test/Syntax/Parsing/SimpleProgramsParsingTests.cs [hunk 6]
```diff
@@ -1627,5 +1603,146 @@ public void TupleUnsupportedInUsingStatement()
...
+                        N(SyntaxKind.IdentifierToken, "e");
+                    }
+                }
+                N(SyntaxKind.EndOfFileToken);
+            }
+            EOF();
+        }
+
+        [Fact]
+        public void TestSkippedTest()
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Syntax/Parsing/SimpleProgramsParsingTests.cs]: I don't understand this test name

Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Syntax/Parsing/SimpleProgramsParsingTests.cs]: > I don't understand this test name

The scenario is taken from the other test named this way.

---
In reply to: [388549580](https://github.com/dotnet/roslyn/pull/42147#discussion_r388549580) [](ancestors = 388549580)
